### PR TITLE
Task payload rework

### PR DESCRIPTION
**Pull request by amdrexu**
In this change, we rework task payload lowering. Previously, task payload is lowered in frontend with a special path in SPIR-V reader and LLPC frontend lower. This is very awkward. Now, we just follow the discussion of
https://github.com/GPUOpen-Drivers/llpc/pull/2354#pullrequestreview-1359385375.

We defer the lowering of task payload to PatchBufferOp. We just transform the task payload buffer descriptor in MeshTaskShader lowering. In the frontend, a virtual call is made: lgc.task.payload.ptr to represent the getter of task payload pointer.

By doing so, many processing codes of task payload in both frontend and in LGC could be removed to share the processing of regular buffer operations since task payload is just a special memory buffer.

### CODE REVIEW

**Comment by amdrexu** [comment]
This is to address the issue: https://github.com/GPUOpen-Drivers/llpc/issues/2373

**Comment by amdrexu** [comment]
This is a real CTS: dEQP-VK.mesh_shader.ext.misc.payload_read
[payload.zip](https://github.com/GPUOpen-Drivers/llpc/files/12158090/payload.zip)

**Comment by amdvlk-admin** [comment]
### Test summary for commit 319fb88db03484ca6648f5bdf64f56ce8307d0f1
<details><summary>CTS tests (Failed: 0/211879)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35146/70631 (49.8%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35485/70631 (50.2%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39025/70632 (55.3%)</li>
<li>Failed: 0/70632 (0.0%)</li>
<li>Not Supported: 31607/70632 (44.7%)</li><li>Warnings: 0/70632 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39057/70616 (55.3%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31559/70616 (44.7%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLowerGlobal.cpp [hunk 1]
```diff
@@ -2630,6 +2084,68 @@ void SpirvLowerGlobal::lowerAliasedVal() {
...
+    // Record of all the functions that our global is used within.
+    SmallSet<Function *, 4> funcsUsedIn;
+
+    for (User *const user : global.users()) {
+      if (Instruction *const inst = dyn_cast<Instruction>(user))
+        funcsUsedIn.insert(inst->getFunction());
+    }
+
+    for (Function *const func : funcsUsedIn) {
+      m_builder->SetInsertPointPastAllocas(func);
```
Comment (nhaehnle) [inline llpc/lower/llpcSpirvLowerGlobal.cpp:2116]: Should we start to adopt `llvm::convertUsersOfConstantsToInstructions` for these kinds of things?

**Review by nhaehnle — COMMENTED** [review]
This is a really nice cleanup, thanks!

I'm a bit concerned that the barrier in the example ends up including cache flush instructions which should not be needed. But I don't think that's related to this particular change.

File: lgc/patch/MeshTaskShader.cpp [hunk 2]
```diff
@@ -1687,76 +1409,72 @@ void MeshTaskShader::lowerMeshShaderBody(BasicBlock *apiMeshEntryBlock, BasicBlo
...
-        removedCalls.push_back(call);
+    for (auto user : func.users()) {
+      CallInst *const call = cast<CallInst>(user);
+
+      if (call->getFunction() != entryPoint)
+        continue; // Not belong to mesh shader
+
+      m_builder.SetInsertPoint(call);
+
+      if (isa<TaskPayloadPtrOp>(call)) {
```
[the remark this answers is not in the document]
Comment (amdrexu) [reply lgc/patch/MeshTaskShader.cpp]: Good idea. I use dialect visitor. I will change all other ops to dialect in follow-up changes.

File: llpc/lower/llpcSpirvLowerGlobal.cpp (hunk 1 above)
Comment (amdrexu) [reply llpc/lower/llpcSpirvLowerGlobal.cpp]: This is a good method. I didn't notice it.

**Comment by amdrexu** [comment]
> I'm a bit concerned that the barrier in the example ends up including cache flush instructions which should not be needed. But I don't think that's related to this particular change.

Do you mean the buffer invalidation instructions are not necessary? They are added by this code piece before we handle `EmitMeshTasks` with an explicit `Fence` instruction:
```
  {
    m_builder.SetInsertPoint(checkEmitMeshsBlock->getTerminator());

    if (m_accessTaskPayload) {
      // Make sure the task payload read/write access is completed
      m_builder.CreateFence(AtomicOrdering::Release, SyncScope::System);
      createBarrier();
    }

    auto firstThreadInSubgroup = m_builder.CreateICmpEQ(m_waveThreadInfo.threadIdInSubgroup, m_builder.getInt32(0));
    m_builder.CreateCondBr(firstThreadInSubgroup, emitMeshsBlock, endEmitMeshsBlock);
    checkEmitMeshsBlock->getTerminator()->eraseFromParent(); // Remove old terminator
  }
```

**Comment by nhaehnle** [comment]
I meant the `barrier()` in the GLSL which becomes buffer invalidate instruction that shouldn't be necessary because the payload stores and loads use GLC anyway.

**Comment by amdrexu** [comment]
> I meant the `barrier()` in the GLSL which becomes buffer invalidate instruction that shouldn't be necessary because the payload stores and loads use GLC anyway.

I see. The fence is added by SPIR-V translator. The handling of memory semantics causes this problem as you said.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5240d05bb3ec277bc865a0dac0d27f59764e740c
<details><summary>CTS tests (Failed: 0/211879)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35146/70631 (49.8%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35485/70631 (50.2%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39025/70632 (55.3%)</li>
<li>Failed: 0/70632 (0.0%)</li>
<li>Not Supported: 31607/70632 (44.7%)</li><li>Warnings: 0/70632 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39057/70616 (55.3%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31559/70616 (44.7%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**